### PR TITLE
3_lambda++ revision, part 1

### PR DESCRIPTION
**Pull request by grosu**
I'd like to merge this before @andreistefanescu merges the `--superstrict`, if possible.  Then I want to work on part 2, which requires a correctly working --search.

@cos , @laurayuwen : can you guys take a quick look and merge?

### CODE REVIEW

**Comment by grosu** [comment]
@laurayuwen , can you please revise this asap?

**Comment by laurayuwen** [comment]
@grosu I am just back from TA training, will look at it right now.

File: k-distribution/tutorial/1_k/1_lambda/lesson_1/tests/config.xml [hunk 1]
```diff
@@ -7,8 +7,7 @@
     definition="lambda.k"
     programs="."
     extension="lambda"
-    results="."
-    skip="pdf">
+    results=".">
     <kompile-option name="--syntax-module" value="LAMBDA" />
```
Comment (laurayuwen) [inline k-distribution/tutorial/1_k/1_lambda/lesson_1/tests/config.xml:11]: Actually, when I went through all the tests, I find that some simple definitions can generate pdf successfully. Haven't really checked the pdf, but it does not jump out as failure during ktest.

File: k-distribution/tutorial/1_k/3_lambda++/lesson_2/README.md [hunk 2]
```diff
@@ -64,26 +76,40 @@ considering how much it does.  Note also that everything that this K rule
...
+terms in the rule, that is, it is not matched when the rule is applied.
+Instead, a fresh natural number is generated each time the rule is applied.
+In K, we can define syntactic categories which have the capability to
+generate fresh elements like above, using unbound variables whose name starts
+with a `!`.  The details of how to do that are beyond the scope of this
+tutorial (see Tutorial 6).  All we need to know here is that an arbitrary
+fresh element of that syntactic category is generated each time the rule
+is applied.  We cannot rely on the particular name, value or structure of
+the generated element, because that can change with the next version of the
+K tool, or even from execution to execution with the same version.  All you
```
Comment (laurayuwen) [inline k-distribution/tutorial/1_k/3_lambda++/lesson_2/README.md]: This is confusing me, what is the structure of the generated element to do with different K versions? We generate a fresh element of the designated syntactic category/type, what is the structure here?

Comment (laurayuwen) [inline k-distribution/tutorial/1_k/3_lambda++/lesson_2/README.md]: lgtm, just not clear with this point.

Comment (grosu) [inline k-distribution/tutorial/1_k/3_lambda++/lesson_2/README.md]: `3` vs `symNat(3)` vs `symNat(sNat(sNat(sNat(zeroNat))))` vs `syntax ::= "v3"`, etc.  I'll try to improve the text, let me know if you have any suggestions.

File: k-distribution/tutorial/1_k/3_lambda++/lesson_1/exercises/callCC/README.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<!-- Copyright (c) 2013-2016 K Team. All Rights Reserved. -->
+
+Define a variant of `callcc`, say `callCC`, which never returns to the
```
Comment (cos) [inline k-distribution/tutorial/1_k/3_lambda++/lesson_1/exercises/callCC/README.md:3]: differentiation by capitalization may be a bit confusing -- not that important tho

File: k-distribution/tutorial/1_k/3_lambda++/lesson_2/README.md [hunk 4]
```diff
@@ -28,35 +29,46 @@ the program execution starts with an empty environment and an empty store.
...
-strictness attributes, your should also define some K results.
+Therefore, unlike in the substitution-based definitions of LAMBDA and
+LAMBDA++, neither the lambda abstractions nor the identifiers are values
+anymore here, because they both evaluate further: lambda abstractions to
+closures and identifiers to their values in the store.  In fact, the only
+values at this moment are the closures, and they are purely semantic entities,
+which cannot be used explicitly in programs.  That's why we modified the
+original syntax of the language to include no `Val` syntactic category
+anymore, and that's why we need to add closures as values now; we do it
+same like before, adding a `Val` syntactic category which is subsorted
```
Comment (cos) [inline k-distribution/tutorial/1_k/3_lambda++/lesson_2/README.md]: "we do it same like before, adding a ..." -> "same as before, we add a ..."

File: k-distribution/tutorial/1_k/3_lambda++/lesson_2/README.md [hunk 5]
```diff
@@ -28,35 +29,46 @@ the program execution starts with an empty environment and an empty store.
...
+to `KResult`.  In general, whenever you have any strictness attributes,
+your should also define some K results.
 
 Invoking a closure is a bit more involved than the substitution-based
-beta-reduction: we need to switch to closure's environment, then create a new
-binding for closure's parameter, then evaluate the closure's body, and
-then switch back to caller's environment, which needs to be stored somewhere
-in the meanwhile.  I prefer to do all these with one rule:
+beta-reduction: we need to switch to closure's environment, then create a
```
Comment (cos) [inline k-distribution/tutorial/1_k/3_lambda++/lesson_2/README.md]: the closure's

File: k-distribution/tutorial/1_k/3_lambda++/lesson_2/README.md [hunk 6]
```diff
@@ -28,35 +29,46 @@ the program execution starts with an empty environment and an empty store.
...
+new, or fresh, binding for closure's parameter to the value passed to the
```
Comment (cos) [inline k-distribution/tutorial/1_k/3_lambda++/lesson_2/README.md]: same

File: k-distribution/tutorial/1_k/3_lambda++/lesson_2/README.md [hunk 7]
```diff
@@ -28,35 +29,46 @@ the program execution starts with an empty environment and an empty store.
...
+closure, then evaluate the closure's body, and then switch back to caller's
```
Comment (cos) [inline k-distribution/tutorial/1_k/3_lambda++/lesson_2/README.md]: the caller's

File: k-distribution/tutorial/1_k/3_lambda++/lesson_2/README.md [hunk 8]
```diff
@@ -28,35 +29,46 @@ the program execution starts with an empty environment and an empty store.
...
+         <store>... .Map => (!N:Int |-> V) ...</store>
 
 Therefore, we atomically do all the following:
 
-- switch the computation to closure's body followed by a
-caller-environment-recovery task `env(Rho')`,
-- generate a fresh location `N`, bind `X` to `N` in closure's environment
-and switch the current environment `Rho'` to that one,
-- write `V` at location `N`.
+- switch the computation to closure's body, `E`, followed by a
```
Comment (cos) [inline k-distribution/tutorial/1_k/3_lambda++/lesson_2/README.md]: again, "the", but maybe it is a choice to use it without an article

File: k-distribution/tutorial/1_k/3_lambda++/lesson_2/README.md [hunk 9]
```diff
@@ -64,26 +76,40 @@ considering how much it does.  Note also that everything that this K rule
...
-      [structural]
+  rule <k> _:Val ~> (Rho => .) ...</k> <env> _ => Rho </env>
 
-When the preceding expression becomes a value, `env(Rho)` replaces the current
-environment to `Rho` and dissolves itself.  Note that, syntactically, we
-preferred `env(Rho)` to be a K term.  Indeed, it does not make much sense for
-it to be an expression construct.
+When the item preceding the environment recovery task `Rho` in the
+computation becomes a value, replace the current environment with `Rho`
+and dissolves `Rho` from the computation.
```
Comment (cos) [inline k-distribution/tutorial/1_k/3_lambda++/lesson_2/README.md]: "dissolve", i.e., without "-s"

File: k-distribution/tutorial/1_k/1_lambda/lesson_1/tests/config.xml (hunk 1 above)
Comment (grosu) [inline k-distribution/tutorial/1_k/1_lambda/lesson_1/tests/config.xml:11]: I tried it and it looks like kdoc simply does not generate anything, but it does not report an error either.  For uniformity, I'm just removing them from anywhere.  I guess @cos will be happy to add them once everything works with kdoc.

**Comment by grosu** [comment]
Thanks @cos and @laurayuwen , new version fixing everything you suggested pushed.

**Pull request merged**